### PR TITLE
More deadlock fixes

### PR DESCRIPTION
**Pull request by vmx**
This commit just shuffles some code around without doing any functional changes.

For review purpose: I also changes some ordering of things, but I think things are still correct. So you only need to look at the general structure, the code within the iterators didn't change.

### CODE REVIEW

**Comment by cryptonemo** [comment]
I like this re-factor -- it does make the threadpool usage more explicit and fixes the remaining GPU multiexp race w/o changes (aside from a potential lock bug remaining).  I don't know why CI is terribly unhappy though.

**Comment by vmx** [comment]
I also cherry-picked https://github.com/filecoin-project/bellperson/pull/108 which was approved but never merged.

**Comment by cryptonemo** [comment]
Sorry about the force push -- fixed some typos that were bothering me.

File: src/multicore.rs [hunk 1]
```diff
@@ -48,10 +54,51 @@ impl Worker {
         R: Send + 'static,
     {
         let (sender, receiver) = bounded(1);
-        THREAD_POOL.spawn(move || {
-            let res = f();
-            sender.send(res).unwrap();
-        });
+
+        let thread_index = if THREAD_POOL.current_thread_index().is_some() {
```
Comment (vmx) [inline src/multicore.rs:58]: Could be written as:

```rust
let thread_index = THREAD_POOL.current_thread_index().unwrap_or(0);
```

File: src/multicore.rs [hunk 2]
```diff
@@ -48,10 +54,51 @@ impl Worker {
...
+
+        // If the number of spawns requested has exceeded the number
+        // of cores available for processing by some factor (the
+        // default being 4), instead of requesting that we spawn a new
+        // thread, we instead execute the closure in the context of an
+        // install call to help clear the growing work queue and
+        // minimize the chances of memory exhaustion.
+        if previous_count > *WORKER_SPAWN_MAX_COUNT {
+            THREAD_POOL.install(move || {
+                debug!("[{}] switching to install to help clear backlog[current threads {}, threads requested {}]",
```
Comment (vmx) [inline src/multicore.rs:81]: Perhaps this should even be `trace!()` as I'd expect it to create quite a lot of noise.

**Comment by vmx** [comment]
Closing in favour of #150.

**Pull request closed without merging**